### PR TITLE
feat(forms): add support for adding async validators via template

### PR DESCRIPTION
**Pull request by vsavkin**
Example:

```
@Directive({
  selector: '[uniq-login-validator]',
  providers: [provide(NG_ASYNC_VALIDATORS, {useExisting: UniqLoginValidator, multi: true})]
})
class UniqLoginValidator implements Validator {
  validate(c) { return someFunctionReturningPromiseOrObservable(); }
}
```

### CODE REVIEW

File: modules/angular2/src/core/forms/directives/ng_control_name.ts [hunk 1]
```diff
@@ -85,17 +91,16 @@ export class NgControlName extends NgControl implements OnChanges,
...
 
   constructor(@Host() @SkipSelf() parent: ControlContainer,
-              @Optional() @Inject(NG_VALIDATORS) validators:
+              @Optional() @Inject(NG_VALIDATORS) private _validators:
+                  /* Array<Validator|Function> */ any[],
+              @Optional() @Inject(NG_ASYNC_VALIDATORS) private _asyncValidators:
                   /* Array<Validator|Function> */ any[],
               @Optional() @Inject(NG_VALUE_ACCESSOR) valueAccessors: ControlValueAccessor[]) {
     super();
     this._parent = parent;
```
Comment (vicb) [inline modules/angular2/src/core/forms/directives/ng_control_name.ts]: nit: could you update the ctor signature and drop this line ?

File: modules/angular2/src/core/forms/form_builder.ts [hunk 2]
```diff
@@ -34,7 +34,7 @@ import * as modelModule from './model';
  *       login: ["", Validators.required],
  *       passwordRetry: builder.group({
  *         password: ["", Validators.required],
- *         passwordConfirmation: ["", Validators.required]
+ *         passwordConfirmation: ["", Validators.required, asyncValidation]
```
Comment (vicb) [inline modules/angular2/src/core/forms/form_builder.ts]: Validation -> Validator ?

File: modules/angular2/src/core/forms/validators.ts [hunk 3]
```diff
@@ -76,14 +78,33 @@ export class Validators {
...
+    if (presentValidators.length == 0) return null;
+
+    return function(control: modelModule.AbstractControl) {
+      return PromiseWrapper.all(_executeValidators(control, presentValidators)).then(_mergeErrors);
+    };
+  }
+}
+
+function _executeValidators(control: modelModule.AbstractControl, validators: Function[]): any[] {
+  return validators.map(v => isPresent(v) ? v(control) : null);
```
Comment (vicb) [inline modules/angular2/src/core/forms/validators.ts]: do we need `isPresent()`  here ? (the list should be filtered already)

File: modules/angular2/test/core/forms/directives_spec.ts [hunk 4]
```diff
@@ -54,6 +56,19 @@ class CustomValidatorDirective implements Validator {
   validate(c: Control): {[key: string]: any} { return {"custom": true}; }
 }
 
+function asyncValidator(expected, timeout = 0) {
+  return (c) => {
+    var completer = PromiseWrapper.completer();
+    var res = c.value != expected ? {"async": true} : null;
+    if (timeout == 0) {
+      PromiseWrapper.scheduleMicrotask(() => { completer.resolve(res); });
```
Comment (vicb) [inline modules/angular2/test/core/forms/directives_spec.ts]: q: why do we need a microtask here ?

**Comment by vicb** [comment]
LGTM

File: modules/angular2/src/core/forms/directives/ng_control_name.ts (hunk 1 above)
Comment (vsavkin) [inline modules/angular2/src/core/forms/directives/ng_control_name.ts]: Fixed

File: modules/angular2/src/core/forms/form_builder.ts (hunk 2 above)
Comment (vsavkin) [inline modules/angular2/src/core/forms/form_builder.ts]: Fixed

File: modules/angular2/src/core/forms/validators.ts (hunk 3 above)
Comment (vsavkin) [inline modules/angular2/src/core/forms/validators.ts]: Fixed

**Pull request closed without merging**